Blog post summarizing 1.36 release cycle (#564)
k3s-io/docs: blog/2026-05-27-K3s-1.36-release.md

<!-- truncate -->

## Key Features and Improvements ✨

### Core Storage Evolution: etcd 3.6 & SQLite Upgrades 💾
We have officially upgraded our bundled etcd to the v3.6 series. While this brings large performance improvements and memory optimizations upstream, it introduces a critically important caveat for administrators running High Availability (HA) clusters: etcd v3.6 changes the data directory structure, which prevents a clean downgrade to v3.5 once your cluster has updated. If you ever need to rollback a v1.36 cluster to v1.35, you must follow specific manual snapshot restoration and node cleaning procedures. Be sure to review our dedicated migration guide before upgrading production environments.

### Revolutionizing Builds with the Nix Snapshotter ❄️
In an exciting move for reproducibility and declarative environments, K3s now includes native support for the Nix snapshotter plugin. This integration allows users to leverage Nix store paths directly as container layers, opening the door for blazingly fast image provisioning and native NixOS container deployments within K3s clusters.

### Streamlined Runtimes: Default Imports to Containerd Base Templates 🏗️

> Customizing your container runtime configuration just got much cleaner. K3s now supports default imports inside our containerd base configuration templates. This architectural change allows administrators to modularly inject runtime configurations and snippets without having to rewrite or completely overwrite the entire default`config.toml.tmpl` file.

Reviewer: ```suggestion
Customizing your container runtime configuration just got much easier. K3s now supports importing containerd configuration drop-ins. This architectural change allows administrators to modularly add configuration snippets that add to or replace configuration from the base file without having to extend or replace the entire default configuration template.
```

Author: yes, makes sense

### Dynamic Security: Secrets Encryption for Existing Clusters 🔐

> Security is never static, and neither should your cluster configuration be. Historically, enabling at-rest Secrets Encryption required setting it up during the initial cluster bootstrapping phase. With v1.36, you can now enable secrets encryption on existing, running clusters without needing to recreate them, making it vastly easier to harden legacy deployments.

Reviewer: ```suggestion
Security is never static, and neither should your cluster configuration be. Historically, enabling at-rest Secrets Encryption required setting it up during the initial cluster bootstrapping phase. With v1.36, you can now enable secrets encryption on existing clusters without needing to recreate them, making it vastly easier to harden legacy deployments. You can also switch between encryption providers - both `aescbc` and `secretbox` are natively supported.
```

---
title: Kubernetes v1.36 is out!
description: A deep dive into the latest K3s features, security enhancements, and community milestones
authors: manuelbuil
hide_table_of_contents: true
---

Kubernetes 1.36 has officially landed! Following our continuous commitment to providing a lightweight yet production-grade Kubernetes distribution, the K3s team and community have put together another incredible milestone release. 🥳

<!-- truncate -->

> This cycle, we focused on strengthening our core infrastructure—bringing major storage innovations, highly requested security enhancements, and empowering advanced workloads like AI and machine learning via modern container runtime features. Here is what's new in the world of K3s.

Reviewer 2: This is way to "high-octane" for what is a medium feature release.

```suggestion
This cycle, we focused on core infrastructure improvements: a new storage backend, highly requested security enhancements, and updated container runtime features. Here is what's new in the world of K3s.
```

Reviewer 2: Also, maybe just combine this with the intro paragraph above. There is no need to have two "intro/exciting" paragraphs.

Author: Gemini got too enthusiastic. Let's just do one paragraph. I also don't think there is a lot of "major innovations"

## Key Features and Improvements ✨

### Core Storage Evolution: etcd 3.6 & SQLite Upgrades 💾

> We have officially upgraded our bundled etcd to the v3.6 series. While this brings massive performance improvements and optimizations upstream, it introduces a critically important caveat for administrators running High Availability (HA) clusters: etcd v3.6 changes the data directory structure, which prevents a clean downgrade to v3.5 once your cluster has updated. If you ever need to rollback a v1.36 cluster to v1.35, you must follow specific manual snapshot restoration and node cleaning procedures. Be sure to review our dedicated migration guide before upgrading production environments.

Reviewer 2: Throughput increased by 10%... that's not "massive"  IMO, but maybe I'm way off base in the world of DBs??

```suggestion
We have officially upgraded our bundled etcd to the v3.6 series. While this brings large performance improvements and memory optimizations upstream, it introduces a critically important caveat for administrators running High Availability (HA) clusters: etcd v3.6 changes the data directory structure, which prevents a clean downgrade to v3.5 once your cluster has updated. If you ever need to rollback a v1.36 cluster to v1.35, you must follow specific manual snapshot restoration and node cleaning procedures. Be sure to review our dedicated migration guide before upgrading production environments.
```

### Revolutionizing Builds with the Nix Snapshotter ❄️
In an exciting move for reproducibility and declarative environments, K3s now includes native support for the Nix snapshotter plugin. This integration allows users to leverage Nix store paths directly as container layers, opening the door for blazingly fast image provisioning and native NixOS container deployments within K3s clusters.

### Streamlined Runtimes: Default Imports to Containerd Base Templates 🏗️

> Customizing your container runtime configuration just got much cleaner. K3s now supports default imports inside our containerd base configuration templates. This architectural change allows administrators to modularly inject runtime configurations and snippets without having to rewrite or completely overwrite the entire default config.toml.tmpl file.

Reviewer 2: ```suggestion
Customizing your container runtime configuration just got much cleaner. K3s now supports default imports inside our containerd base configuration templates. This architectural change allows administrators to modularly inject runtime configurations and snippets without having to rewrite or completely overwrite the entire default `config.toml.tmpl` file.
```

### Dynamic Security: Secrets Encryption for Existing Clusters 🔐

> Security is never static, and neither should your cluster configuration be. Historically, enabling at-rest Secrets Encryption required setting it up during the initial cluster bootstrapping phase. With v1.36, you can now seamlessly enable secrets encryption on existing, running clusters without needing to recreate them, making it vastly easier to harden legacy deployments.

Reviewer 2: Not seamless, as you still need to restart your cluster servers several times. 
```suggestion
Security is never static, and neither should your cluster configuration be. Historically, enabling at-rest Secrets Encryption required setting it up during the initial cluster bootstrapping phase. With v1.36, you can now  enable secrets encryption on existing, running clusters without needing to recreate them, making it vastly easier to harden legacy deployments.
```

Author: good point

### AI & Hardware Acceleration: Containerd 2.2 with NRI Support 🚀

> We have bumped our embedded container runtime to Containerd 2.2, which introduces out-of-the-box support for the Node Resource Interface (NRI). This is a game-changer for clusters running edge AI and heavy GPU workloads, enabling cleaner integration with the NVIDIA Container Toolkit and highly dynamic, runtime-level device management.

Reviewer 2: ```suggestion
We have bumped our embedded container runtime to Containerd 2.2, which introduces out-of-the-box support for the Node Resource Interface (NRI). This is a game-changer for clusters running edge AI and heavy GPU workloads, enabling cleaner integration with the NVIDIA Container Toolkit and runtime-level device management.
```
